toolchain/sem_ir/diagnostic_loc_converter.h

```diff
@@ -0,0 +1,69 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_SEM_IR_DIAGNOSTIC_LOC_CONVERTER_H_
+#define CARBON_TOOLCHAIN_SEM_IR_DIAGNOSTIC_LOC_CONVERTER_H_
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "toolchain/diagnostics/diagnostic_emitter.h"
+#include "toolchain/parse/tree_and_subtrees.h"
+#include "toolchain/sem_ir/absolute_node_id.h"
+#include "toolchain/sem_ir/file.h"
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::SemIR {
+
+class DiagnosticLocConverter {
+ public:
+  // Information about an import within which the location was found.
+  struct ImportLoc {
+    Diagnostics::Loc loc;
+    // TODO: Include the name of the imported library in this information so it
+    // can be included in the diagnostic.
+  };
+
+  // Information about a location that has been converted from a LocId to a
+  // diagnostic location.
+  struct LocAndImports {
+    llvm::SmallVector<ImportLoc> imports;
+    Diagnostics::ConvertedLoc loc;
+  };
+
+  explicit DiagnosticLocConverter(
+      llvm::ArrayRef<Parse::GetTreeAndSubtreesFn> tree_and_subtrees_getters,
+      const File* sem_ir)
+      : tree_and_subtrees_getters_(tree_and_subtrees_getters),
+        sem_ir_(sem_ir) {}
+
+  // Converts the given location into a sequence of import locations and a final
+  // diagnostic location.
+  auto Convert(LocId loc_id, bool is_token_only) const -> LocAndImports;
+
+  // Converts an `absolute_node_id` in either a Carbon file or C++ import to a
+  // diagnostic location.
+  auto Convert(AbsoluteNodeId absolute_node_id, bool token_only) const
```

Comment:
While you're moving this, can you fix is_token_only vs token_only to use the same name?